main/main.c

```diff
@@ -19,18 +19,70 @@
 #include <stdio.h>
 #include <stdbool.h>
 #include <signal.h>
+#include <time.h>
+#include <inttypes.h>
 
 #include <rte_eal.h>
 #include <rte_log.h>
 #include <rte_common.h>
 #include <rte_launch.h>
+#include <rte_cycles.h>
 
+#include "gatekeeper_main.h"
 #include "gatekeeper_config.h"
 #include "gatekeeper_net.h"
 
 /* Indicates whether the program needs to exit or not. */
 volatile int exiting = false;
 
+/*
+ * These metrics are system dependent, and 
+ * initialized via time_resolution_init() function.
+ */
+uint64_t cycles_per_sec;
+uint64_t picosec_per_cycle;
+
+/* Obtain the system time resolution. */
+static int
+time_resolution_init(void)
+{
+	int ret;
+	uint64_t diff_ns;
+	uint64_t cycles;
+	uint64_t tsc_start;
+	struct timespec tp_start;
+
+	tsc_start = rte_rdtsc();
+	ret = clock_gettime(CLOCK_MONOTONIC_RAW, &tp_start);
+	if (ret < 0)
+		return ret;
+
+	while (1) {
+		struct timespec tp_now;
+
+		ret = clock_gettime(CLOCK_MONOTONIC_RAW, &tp_now);
+		if (ret < 0)
+			return ret;
+
+		diff_ns = (uint64_t)(tp_now.tv_sec - tp_start.tv_sec) * 1000000000UL 
+				+ (uint64_t)(tp_now.tv_nsec - tp_start.tv_nsec);
+
+		if (diff_ns >= 1000000000UL) {
+			cycles = rte_rdtsc() - tsc_start;
```

Comment:
We need `tsc_now = rte_rdtsc();` immediately after `ret = clock_gettime(CLOCK_MONOTONIC_RAW, &tp_now);` because we currently adding the cycles after `clock_gettime()` returns and here. This should be a small accuracy improvement.